http/template/tag_packer.go

```diff
@@ -60,3 +63,41 @@ func TagPacker(env trails.Environment, filesys fs.FS) func(string, bool) html.HT
 		}
 	}
 }
+
+// AssetURI encloses the environment and filesystem so when called executing a template,
+// emits valid URI for client side static and bundled assets.
+//
+// TODO:
+//
+// - configurable OriginHost would allow for support of a CDN and custom Vite server origin
+func AssetURI(env trails.Environment, filesys fs.FS) func(string) string {
+	if filesys == nil {
+		filesys = os.DirFS(".")
+	}
+
+	return func(assetPath string) string {
+		switch {
+		case env.IsTesting():
+			return ""
+
+		case env.IsDevelopment():
+			return fmt.Sprintf("http://localhost:8080/%s/%s", assetsBase, assetPath)
+
+		default:
+			// match hashed files bundled by Vite
+			filename := strings.TrimSuffix(assetPath, filepath.Ext(assetPath))
+			fileExt := filepath.Ext(assetPath)
+
+			// Note: where assetPath = assets/GetDashboard.js
+			// glob = client/dist/assets/GetDashboard-*.js
+			glob := fmt.Sprintf("%s/%s-*%s", assetsBase, filename, fileExt)
+			matches, err := fs.Glob(filesys, glob)
+
+			if errors.Is(err, path.ErrBadPattern) || len(matches) == 0 {
```

Comment:
This one is tricky, but I feel like if we don't match the pattern when looking for a hashed asset, we should leave the path unchanged.  This lets the logs show the request path if it 404's.  I say if, because a static asset will not match this glob pattern, since they are not hashed and should resolve as expected.

The alternative to the glob strategy is to consume a manifest, which has some pros/cons, but to leverage it should probably be geared more towards the "optimization" it provides by getting all of an entry points dependencies on to the page with `<script module="preload">`.

Reply:
wrt globbing, it makes me wonder:
1. How brittle is this? I'm concerned about footguns. i.e. What are the odds of matching the wrong file because somebody didn't name something _just right_?
2. Probably not an issue in our world, but I'd be curious to know what kind of perf hit all this globbing causes. Maybe some memoization is worth exploring here?
3. I'd be interested to hear a little more about your thoughts on utilizing a manifest.

Reply:
@seedifferently 

1.  It's reasonably resilient based on current experience, but has the potential to become problematic.  Trails matches against the single main stylesheet and the JS entrypoint files for every page load.  Since we keep all entrypoint files in a single src directory, you can't create a naming collision, but because Vite/Rollup will create chunks for composable and component files as well, one of those could cause a collision.   ex: We have numerous `Main.vue` files that are likely to be chunked to `Main-xxxxx.js`, if you created an entrypoint file amed `Main.ts` - you'll have a hard time getting the right file to match.
2. It definitely could be a perf issue at some point, but I'm not terrible concerned right now. We're only making 2-3 match attempts: `style-xxxxxx.css`, `GetEntrypoint-xxxxxx.js` and potentially `favicon.ico`.  Consuming a manifest at bootstrap and storing it in memory for an indexed lookup would certainly resolve any issues there.
3. I think the glob match is nice in the sense that it doesn't rely on the toolchain creating a manifest.  Vite does it out of the box with a single config option, but that wasn't true with Webpack the last time I used it, and I'm uncertain what other alternatives like Parcel.js are doing.  That being said, I've assumed we'd either need or want the benefits eventually, once we've decided they outweigh the added complexity.  One way I've leveraged the manifest in the past is to use its presence in the file system as an indicator of the mode my client side is in.  Since a client side dev server mode is not necessarily the same as being in ENV=development, finding the asset paths based on the existence of a `manifest.json` can at time be very helpful.

The major advantages to consuming a manifest:

- separates client side mode from server side ENV
- faster and more resilient asset matching for entrypoints
- we can drop the conditional <script> in the template, as the manifest key is typically the `source` path
- (optional): The manifest provides a list of chunks that are dependencies of your entrypoint, so we can revist something like TagPacker to spit out all the client side assets we know the page is going to request asynchronously and tag them with the preload attribute.  This would include stylesheets as chunks as well, but based on our setup I'm not certain there's much of a win there.

Cache busting:

I have not reconfirmed this, but the standard for years with assets that go through the build pipeline has been to hash the chunked files contents.  This keeps you from busting the cache on files that don't change often, like vendor dependencies in node_modules, while also ensuring that chunks can be optimized as needed on each build.

A static asset would require a filename change to ensure a busted cache, since it is not hashed and is served by the same handler.

Reply:
ok great thanks for filling in the gaps for me. I'm 👍 with this. My only other bit of feedback might be to log a warning if `len(matches) > 1` so there's some visibility around that scenario should it ever happen.